drivers: led: led_gpio: Add support of blink

anobli opened the pull request:
The led_api_blink is quite convenient to use.
This adds its support to the led_gpio driver.

anobli committed:
drivers: led: led_gpio: Add support of blink

The led_api_blink is quite convenient to use.
This adds its support to the led_gpio driver.

Signed-off-by: Alexandre Bailon <abailon@baylibre.com>

simonguinot reviewed (changes requested):
Hi @anobli,

Please see https://github.com/zephyrproject-rtos/zephyr/pull/57791.

henrikbrixandersen commented:
> Hi @anobli,
> 
> 
> 
> Please see https://github.com/zephyrproject-rtos/zephyr/pull/57791.

I agree that a general, optional fallback would be better.

anobli commented:
OK, indeed, this makes sense. I will think about to another implementation more adapted to the fallback.

anobli committed:
drivers: led: Provide a fallback implementation of blink API

The LED blink API i squite useful but this is not supported by all
the LED controller.
This provides a software implementation of blink that relies on timer
to blink a LED.
This could be enabled using Kconfig, as well the maximum number of LED
that could blink at the same time using blink fallback.
This does not requires any changes to driver to be used.

Signed-off-by: Alexandre Bailon <abailon@baylibre.com>

soburi commented on include/zephyr/drivers/led.h:
```diff
@@ -145,7 +156,7 @@ static inline int z_impl_led_blink(const struct device *dev, uint32_t led,
 
 	if (api->blink == NULL) {
-		return -ENOSYS;
+		return led_blink_fallback(dev, led, delay_on, delay_off);
```
I think it would be better to put this in an ifdef and implement it so that it keeps clean when it's disabled.

soburi commented on include/zephyr/drivers/led.h:
```diff
@@ -308,6 +319,7 @@ static inline int z_impl_led_on(const struct device *dev, uint32_t led)
 	const struct led_driver_api *api =
 		(const struct led_driver_api *)dev->api;
 
+	led_blink_fallback(dev, led, 0, 0);
```
ditto

soburi commented on include/zephyr/drivers/led.h:
```diff
@@ -327,6 +339,7 @@ static inline int z_impl_led_off(const struct device *dev, uint32_t led)
 	const struct led_driver_api *api =
 		(const struct led_driver_api *)dev->api;
 
+	led_blink_fallback(dev, led, 0, 0);
```
ditto

soburi commented on drivers/led/led_gpio.c:
```diff
@@ -24,6 +24,65 @@ struct led_gpio_config {
+	const struct gpio_dt_spec *led_gpio = &config->led[led];
+
+	data->blink_state ^= 1 << led;
+	gpio_pin_set_dt(led_gpio, data->blink_state & (1 << led));
```
Can this be replaced with `led_on` to work with other LED devices?

anobli replied:
I left it by mistake. Actually, all the change made in led_gpio.c should be reverted.

anobli commented:
> Hi @anobli,
> 
> Maybe you should close this PR, and take the time to write a proper driver for software blinking ? And once ready you could open a new PR, with the right title and a clean history ?
> 
> In addition, are you sure that timer is the right tool to implement software blinking ? There are various backends behind LEDs: GPIO, SPI, I2C, PWM, and many others. Some of them may even sleep during I/O. Is that allowed in timer context in Zephyr ? On all platforms ? In Linux for example, timers are indeed used for software blinking. But if the LED end-driver doesn't provide the `brightness_set` API method which is guaranteed to not sleep, then a work queue is used, to provide a user sleep-friendly context.
> 
> This software blinking driver will be quite an important one. So please take the time to make it great and optimized.

I will close this PR, and open a new one soon.
Indeed, the timer is not the best choice.
I will use a delayable work that doesn't seem to have any big memory overhead compared to timer and deal correctly with sleep context.

anobli commented:
Closing this PR since the scope has changed.
Thanks everyone for the review you made!
